jax/core.py

```diff
@@ -1576,7 +1576,7 @@ def extend_axis_env(axis_name, size: int, tag: Any):
       yield
     finally:
       frame_ = thread_local_state.trace_state.axis_env.pop()
-      assert frame is frame_
+    assert frame is frame_  # Only runs if there was was no exception
```

Comment:
Hmm don't we want to run this check even if there is an exception, to ensure that our cleanup procedure is happening as expected? O maybe that obscures the underlying exception or something?

Reply:
Yeah it has obscured another problem making it really hard to debug.